File: advance/src/eu/advance/logistics/flow/engine/api/AdvanceKeyStore.java
/*
 * Copyright 2010-2012 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */

package eu.advance.logistics.flow.engine.api;

import hu.akarnokd.reactive4java.base.Func0;
import eu.advance.logistics.flow.engine.xml.typesystem.XElement;
import eu.advance.logistics.flow.engine.xml.typesystem.XSerializable;

/**
 * A record representing information about local or remote key stores.
 * @author karnokd, 2011.09.20.
 */
public class AdvanceKeyStore extends AdvanceCreateModifyInfo 
implements XSerializable, HasPassword, Copyable<AdvanceKeyStore> {
	/** The key store name. */
	public String name;
	/** The key store location on disk. */
	public String location;
	/** 
	 * The password.
	 * <p>An empty password should be an empty {@code char} array. To keep
	 * the current password, use {@code null}.</p>
	 * <p>Note that passwords are never returned from the 
	 * control API calls and are always {@code null}.</p> 
	 * @return the password. 
	 */
	private char[] password;
	/** The function to create a new instance of this class. */
	public static final Func0<AdvanceKeyStore> CREATOR = new Func0<AdvanceKeyStore>() {
		@Override
		public AdvanceKeyStore invoke() {
			return new AdvanceKeyStore();
		}
	};
	@Override
	public void load(XElement source) {
		name = source.get("name");
		location = source.get("location");
		password = getPassword(source, "password");
		super.load(source);
	}
	@Override
	public void save(XElement destination) {
		destination.set("name", name);
		destination.set("location", location);
		setPassword(destination, "password", password);
		super.save(destination);
	}
	@Override
	public AdvanceKeyStore copy() {
		AdvanceKeyStore result = new AdvanceKeyStore();
		
		result.name = name;
		result.location = location;
		result.password = password != null ? password.clone() : null;
		
		assignTo(result);
		
		return result;
	}
	@Override
	public char[] password() {
		return password != null ? password.clone() : null;
	}
	@Override
	public void password(char[] newPassword) {
		password = newPassword != null ? newPassword.clone() : null;
	}
}

File: advance-api-cs/AdvanceAPIClient/Classes/AdvanceKeyStore.cs
/*
 * Copyright 2010-2013 The Advance EU 7th Framework project consortium
 *
 * This file is part of Advance.
 *
 * Advance is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Advance is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with Advance.  If not, see
 * <http://www.gnu.org/licenses/>.
 *
 */
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Globalization;
using System.Xml;

using AdvanceAPIClient.Core;
using AdvanceAPIClient.Interfaces;

namespace AdvanceAPIClient.Classes
{
    public class AdvanceKeyStore : AdvanceCreateModifyInfo, IPassword, ICopiable<AdvanceKeyStore>, IIdentifiable<string>
    {
        /// <summary>
        /// name used by blocks to reference this data source
        /// </summary>
        public string Id { get { return this.Name; } }
        public string Name;
        /// <summary>
        /// Location prefix to emulate a working directory.
        /// </summary>
        public string LocationPrefix = "";
        /// <summary>
        /// Key store location on disk.
        /// </summary>
        public string Location;

        /// <summary>
        /// User's password.
        /// <remarks>
        /// Note that passwords are never returned from the 
        /// control API calls and are always null
        /// An empty password should be an char[0] array. To keep
        /// the current password, use null
        /// </remarks>
        /// </summary>
        public char[] Password
        {
            get { return (this.password == null) ? null : (char[])this.password.Clone(); }
            set { this.password = (value == null) ? null : (char[])value.Clone(); }
        }
        private char[] password;

        public AdvanceKeyStore() : base() { }

        /// <summary>
        /// Fill object data from xml node
        /// </summary>
        /// <param name="source">Source Xml</param>
        protected override void LoadFromXmlNode(XmlNode source)
        {
            base.LoadFromXmlNode(source);
            this.Name = GetAttribute(source, "name");
            this.Location = GetAttribute(source, "location", null);
            this.password = GetPassword(source, "password");
        }

        /// <summary>
        /// Fills  object's xml
        /// </summary>
        /// <param name="node">Empty node for the object with name "tagname"</param>
        protected override void FillXmlElement(XmlElement node)
        {
            base.FillXmlElement(node);
            AddAttribute(node, "name", this.Name);
            AddAttribute(node, "location", this.Location);
            SetPassword(node, "password", password);
        }

        public AdvanceKeyStore Copy()
        {
            AdvanceKeyStore result = new AdvanceKeyStore();
            this.AssignTo(result);
            result.Name = this.Name;
            result.password = this.Password;
            result.Location = this.Location;
            result.LocationPrefix = this.LocationPrefix;
            return result;
        }

    }
}

